kratos/includes/geometry_container.h

```diff
@@ -0,0 +1,407 @@
+    ///@name Type Definitions
+    ///@{
+
+    /// Pointer definition of GeometryContainer
+    KRATOS_CLASS_POINTER_DEFINITION(GeometryContainer);
+
+    typedef std::size_t IndexType;
+    typedef std::size_t SizeType;
+
+    /* Geometry Hash Map Container.
+    *  Hash of Id are keys to corresponding intrusive pointer */
+    typedef PointerHashMapSet<
+        TGeometryType,
+        std::hash<std::size_t>,
+        GetGeometryId,
+        typename TGeometryType::Pointer
+    > GeometriesContainerType;
+
+    /// Geometry Iterator
+    typedef typename GeometriesContainerType::iterator GeometryIterator;
+
+    /// Const Geometry Iterator
+    typedef typename GeometriesContainerType::const_iterator GeometryConstantIterator;
+
+    ///@}
+    ///@name Life Cycle
+    ///@{
+
+    /// Default Constructor
+    GeometryContainer()
+        : Flags()
+        , mpGeometries(new GeometriesContainerType())
+    {}
+
+    /// Copy Constructor
+    GeometryContainer(GeometryContainer const& rOther)
+        : Flags(rOther)
+        , mpGeometries(rOther.mpGeometries)
+    {}
+
+    /// Components Constructor
+    GeometryContainer(
+        typename GeometriesContainerType::Pointer NewGeometries)
+        : Flags()
+        , mpGeometries(NewGeometries)
+    {}
+
+    /// Destructor
+    ~GeometryContainer() override
+    {}
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    GeometryContainer Clone()
+    {
+        typename GeometriesContainerType::Pointer p_geometries(new GeometriesContainerType(*mpGeometries));
+
+        return GeometryContainer(p_geometries);
+    }
+
+    void Clear()
+    {
+        Flags::Clear();
+        DataValueContainer::Clear();
+        mpGeometries->clear();
+    }
+
+    ///@}
+    ///@name Geometries
+    ///@{
+
+    /// Return number of geometries stored inside this geometry container
+    SizeType NumberOfGeometries() const
+    {
+        return mpGeometries->size();
+    }
+
+    ///@}
+    ///@name Add Functions
+    ///@{
+
+    /// Adds a geometry to the geometry container.
+    GeometryIterator AddGeometry(typename TGeometryType::Pointer pNewGeometry)
+    {
+        auto i = mpGeometries->find(pNewGeometry->Id());
+        if(i == mpGeometries->end())
+            return mpGeometries->insert(pNewGeometry);
+        else
+        {
+            KRATOS_ERROR << "Geometry with Id: " << pNewGeometry->Id()
+                << " exists already.";
+        }
+    }
+
+    ///@}
+    ///@name Get Functions
+    ///@{
+
+    /// Returns the Geometry::Pointer corresponding to its Id
+    typename TGeometryType::Pointer pGetGeometry(IndexType GeometryId)
+    {
+        auto i = mpGeometries->find(GeometryId);
+        KRATOS_ERROR_IF(i == mpGeometries->end())
+            << " geometry index not found: " << GeometryId << ".";
+        return (i.base()->second);
+    }
+
+    /// Returns the Geometry::Pointer corresponding to its name
+    typename TGeometryType::Pointer pGetGeometry(std::string GeometryName)
```

Comment:
what is the Geometry-Name?

Reply:
GeometryName allows to use strings as identifiers, similiar to id identifiers. You can check the GeometryId PR #5956